include/xtensor/xreducer.hpp

```diff
@@ -37,20 +37,231 @@ namespace xt
+                else
+                {
+                    temp_idx[(std::size_t) i]++;
+                    break;
+                }
+            }
+            return std::make_pair(i == -1,
+                                  std::inner_product(temp_idx.begin(), temp_idx.end(),
+                                                     iter_strides.begin(), ptrdiff_t(0)));
+        };
+
+        auto begin = e.raw_data();
+        auto out = result.raw_data();
+        auto out_begin = result.raw_data();
+
+        ptrdiff_t next_stride = 0;
+
+        std::pair<bool, ptrdiff_t> idx_res(false, 0);
+
+        // Remark: eventually some modifications here to make conditions faster where merge + accumulate is the
+        // same function (e.g. check std::is_same<decltype(merge_fct), decltype(acc_fct)>::value) ...
+
+        auto merge_border = out;
+        bool merge = false;
+        // Decide if going about it row-wise or col-wise
+        if (inner_stride == 1)
+        {
+            while(idx_res.first != true)
+            {
+                // for unknown reasons it's much faster to use a temporary variable and
+                // std::accumulate here -- probably some cache behavior
+                result_type tmp;
+                tmp = init_fct(*begin);
+                tmp = std::accumulate(begin + 1, begin + outer_loop_size, tmp, acc_fct);
+
+                if (merge)
+                {
+                    *out = merge_fct(*out, tmp);
+                }
+                else
+                {
+                    *out = tmp;
+                }
+
+                begin += outer_loop_size;
+
+                idx_res = next_idx();
+                next_stride = idx_res.second;
+                out = out_begin + next_stride;
+
+                if (out > merge_border)
+                {
+                    // looped over once
+                    merge = false;
+                    merge_border = out;
+                }
+                else
+                {
+                    merge = true;
+                }
+            };
+        }
+        else
+        {
+            while(idx_res.first != true)
+            {
+                if (!merge)
+                {
+                    for (std::size_t i = 0; i < inner_loop_size; ++i)
+                    {
+                        *(out + i) = init_fct(*(begin + i));
+                    }
+                }
+                else
+                {
+                    for (std::size_t i = 0; i < inner_loop_size; ++i)
+                    {
+                        *(out + i) = merge_fct(*(out + i), *(begin + i));
+                    }
+                }
+                begin += inner_stride;
+                for (std::size_t i = 1; i < outer_loop_size; ++i)
+                {
+                    for (std::size_t j = 0; j < inner_loop_size; ++j)
+                    {
+                        *(out + j) = acc_fct(*(out + j), *(begin + j));
+                    }
```

Comment:
This can be simplified with
```cpp
std::transform(out, out + inner_loop_size, begin, out, acc_fct);
```

Reply:
too bad the unary std::transform doesn't exist yet! we could totally use it to parallelize everything and add a `execution_policy` tag! :) C++17, here we come!